### PR TITLE
[RSDK-2753] bump-rdk-post-reconfigure-refactor

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-2753)

- Migrate `slamConfig.AttrConfig` -> `slamConfig.Config`
- Change Close to take context
- Remove dependence on `utils.TryClose`
- Migrate `registry.Dependencies` -> `resource.Dependencies`
- Migrate `config.Service` -> `resource.Config`
- Migrate `registry.RegisterService` -> `resource.RegisterService`
- Remove `config.RegisterServiceAttributeMapConverter` in `init()` as RDK now handles that automatically (somehow)
- Migrate `config.ConvertedAttributes.(*slamConfig.AttrConfig)` -> `resource.NativeConfig[*slamConfig.Config](c)`
- Remove `name` from cartographerService struct
- Embed `resource.Named` and `resource.AlwaysRebuild` in cartographerService struct
- Migrate `cartoSvc.name` -> `c.ResourceName().ShortName()`

Depends on: https://github.com/viamrobotics/slam/pull/190

Manually Tested on:
- [X] M1 Mac
- [x] ARM linux

Screenshot 2023-04-26 at 5 17 38 PM

TODO:
- [x] @kkufieta remove compile time dependence on deprecated slam_libraries.go code

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -313,3 +311,5 @@ require (
 	mvdan.cc/unparam v0.0.0-20221223090309-7455f1af531d // indirect
 	nhooyr.io/websocket v1.8.7 // indirect
 )
+
+replace go.viam.com/slam => github.com/nicksanford/slam v0.0.0-20230426173430-e7fdd7eeb63b
```
Comment (nicksanford) [inline go.mod:315]: Remove before merge

File: integration_test.go [hunk 2]
```diff
@@ -83,7 +82,7 @@ func testCartographerInternalState(t *testing.T, svc slam.Service, dataDir strin
 	test.That(t, err, test.ShouldBeNil)
 }
 
-func integrationtestHelperCartographer(t *testing.T, mode slam.Mode) {
+func integrationtestHelperCartographer(t *testing.T, mode slamTesthelper.Mode) {
```
Comment (nicksanford) [inline integration_test.go:85]: @kkufieta this is using the slam_library remnants from the slam repo's test helper

File: viam-cartographer_test.go [hunk 3]
```diff
@@ -56,31 +55,30 @@ func TestNew(t *testing.T) {
...
 		}
 
-		svc, err := testhelper.CreateSLAMService(t, attrCfg, logger, false, testExecutableName)
+		_, err := testhelper.CreateSLAMService(t, attrCfg, logger, false, testExecutableName)
 		test.That(t, err, test.ShouldBeError,
 			errors.New("configuring lidar camera error: 'sensors' must contain only one "+
 				"lidar camera, but is 'sensors: [lidar, one-too-many]'"))
 
 		grpcServer.Stop()
-		test.That(t, utils.TryClose(context.Background(), svc), test.ShouldBeNil)
```
Comment (nicksanford) [inline viam-cartographer_test.go:79]: This close call is being removed as `svc` is nil, `utils.TryClose` was swallowing a panic

File: viam-cartographer_internal_test.go [hunk 4]
```diff
@@ -118,7 +119,7 @@ func TestGetPositionEndpoint(t *testing.T) {
 			pose, componentRef, err := svc.GetPosition(context.Background())
 			test.That(t, err, test.ShouldBeNil)
 			expectedPose := spatialmath.NewPose(
-				r3.Vector{inputPose.X, inputPose.Y, inputPose.Z},
+				r3.Vector{X: inputPose.X, Y: inputPose.Y, Z: inputPose.Z},
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:122]: This fixes a lint issue

**Review by kkufieta — APPROVED** [review]

File: viam-cartographer.go [hunk 5]
```diff
@@ -70,22 +71,13 @@ func init() {
 			)
 		},
 	})
-
-	config.RegisterServiceAttributeMapConverter(
```
Comment (EshaMaharishi) [inline viam-cartographer.go:74]: this moved to https://github.com/viamrobotics/viam-cartographer/pull/62/files/34ca5218fa1f054b1371fb8be31ae81f2ea44431#diff-93098c528cfadf9d4319fd6b3a1eb42ef437c6a805da14f423688a6944a4d912R54

File: viam-cartographer.go [hunk 6]
```diff
@@ -188,8 +180,8 @@ func New(
 
 // cartographerService is the structure of the slam service.
 type cartographerService struct {
-	generic.Unimplemented
-	name              string
+	resource.Named
+	resource.AlwaysRebuild
```
Comment (EshaMaharishi) [inline viam-cartographer.go:184]: TriviallyReconfigurable won't call close, and we need to call close to shut down the data process and carto_grpc_server

**Review by EshaMaharishi — APPROVED** [review]
lgtm!

**Pull request merged**